src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs

```diff
@@ -58,72 +59,138 @@ public async Task StartAsync(CancellationToken cancellationToken = default)
-            CancellationToken combinedCancellationToken = combinedCancellationTokenSource.Token;
+            CancellationTokenSource linkedCts;
+            if (_options.StartupTimeout != TimeSpan.Zero)
+            {
+                CancellationTokenSource cts = new CancellationTokenSource(_options.StartupTimeout);
+                linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cts.Token, cancellationToken, _applicationLifetime.ApplicationStopping);
+            }
+            else
+            {
+                linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _applicationLifetime.ApplicationStopping);
+            }
 
-            await _hostLifetime.WaitForStartAsync(combinedCancellationToken).ConfigureAwait(false);
+            using (linkedCts)
+            {
+                CancellationToken combinedCancellationToken = linkedCts.Token;
 
-            combinedCancellationToken.ThrowIfCancellationRequested();
-            _hostedServices = Services.GetRequiredService<IEnumerable<IHostedService>>();
+                await _hostLifetime.WaitForStartAsync(combinedCancellationToken).ConfigureAwait(false);
 
-            List<Exception> exceptions = new List<Exception>();
+                combinedCancellationToken.ThrowIfCancellationRequested();
+                _hostedServices = Services.GetRequiredService<IEnumerable<IHostedService>>();
+                _startupServices = Services.GetRequiredService<IEnumerable<IHostedServiceStartup>>();
 
-            if (_options.ServicesStartConcurrently)
-            {
-                List<Task> tasks = new List<Task>();
+                List<Exception> exceptions = new List<Exception>();
 
-                foreach (IHostedService hostedService in _hostedServices)
+                if (_options.ServicesStartConcurrently)
                 {
-                    tasks.Add(Task.Run(() => StartAndTryToExecuteAsync(hostedService, combinedCancellationToken), combinedCancellationToken));
-                }
+                    Task groupedTasks;
+                    List<Task> tasks = new List<Task>();
+
+                    // Handle IServiceStartup.
+                    // For this prototype, we run all startup services first and then wait for them to finish before
+                    // running the IHostedService ones. This supports scenarios such as validation and perhaps preheating
+                    // based on requirements.
+                    // We could have two interfaces instead such as such as IServiceBeforeStart and IServiceAtStart to allow for both modes.
+                    foreach (IHostedServiceStartup startupService in _startupServices)
+                    {
+                        tasks.Add(Task.Run(() => startupService.StartupAsync(combinedCancellationToken).ConfigureAwait(false)));
```

Comment:
Why do we call Task.Run?

Reply:
To call all of the handlers concurrently (instead of serially).